torchdynamo/variables/misc.py

```diff
@@ -243,6 +227,84 @@ def set_grad_insts(mode):
 
         return (prologue, epilogue)
 
+    def _call_func(self, tx, initial_value):
+        raise NotImplementedError("call_func called on base")
+
+    def _func_name(self):
+        raise NotImplementedError("_func_name called on base")
+
+    def _initial_value(self):
+        raise NotImplementedError("_initial_value called on base")
+
+
+class GradModeVariable(ContextWrappingVariable):
+    def __init__(self, target_value, initial_value=None, **kwargs):
+        super(GradModeVariable, self).__init__(
+            target_value=target_value, initial_value=initial_value, **kwargs
+        )
+
+    def enter(self, tx):
+        assert self.initial_value == torch.is_grad_enabled()
+        return super(GradModeVariable, self).enter(tx)
+
+    def _call_func(self, tx, value):
+        if self.target_value == self.initial_value:
+            return
+        tx.output.graph.create_node(
+            "call_function", torch._C._set_grad_enabled, (value,), {}
+        ),
+        torch._C._set_grad_enabled(value)
+
+    def _func_name(self):
+        return "_C._set_grad_enabled"
+
+    def _initial_value(self):
+        return torch.is_grad_enabled()
+
+    def fn_name(self):
+        if self.target_value:
+            return "enable_grad"
+        else:
+            return "no_grad"
+
+
+class ProfileRunModeVariable(ContextWrappingVariable):
```

Comment:
I'd suggest `ProfileRecordFunctionVariable`

_record_function_enter doesn't actually change a mode, it is a way of inserting labels into a profile.

Reply:
agreed.